usl_models/usl_models/flood_ml/batch_predictor.py

```diff
@@ -0,0 +1,191 @@
+
+import json
+import urllib.parse
+import logging
+import dataclasses
+import tensorflow as tf
+
+from google.cloud import firestore  # type:ignore[attr-defined]
+from google.cloud import storage  # type:ignore[attr-defined]
+import numpy as np
+
+from usl_models.flood_ml import model
+from usl_models.flood_ml import prediction_dataset
+
+
+@dataclasses.dataclass
+class BatchPredictor:
+    """Runs batch predictions on files stored in GCS.
+
+    Ouputs data based on the specified `output_bucket`, `model_id`,
+    and `run_id`.
+    """
+
+    db: firestore.Client
+    client: storage.Client
+    output_bucket: str
+    model_id: str
+    model: model.FloodModel
+    run_id: str
+    batch_size: int = 1
+
+    def get_prediction_npy_path(
+        self, study_area_id: str, config_id: str, chunk_id: str
+    ) -> str:
+        """Returns the GCS filepath for a prediction npy file."""
+        return (
+            f"{self.run_id}/flood/{self.model_id}/{study_area_id}"
+            + f"/{config_id}/{chunk_id}"
+        )
+
+    def save_prediction_npy(
+        self, study_area_id: str, config_id: str, chunk_id: str, prediction: np.ndarray
+    ) -> None:
+        """Saves a prediction npy file to GCS."""
+        path = self.get_prediction_npy_path(study_area_id, config_id, chunk_id)
+        logging.info(f"Saving chunk to {path} with shape {prediction.shape}")
+        blob = self.client.get_bucket(self.output_bucket).blob(path)
+        with blob.open("wb") as fd:
+            np.save(fd, prediction)
+
+    def load_prediction_npy(
+        self, study_area_id: str, config_id: str, chunk_id: str
+    ) -> np.ndarray:
+        """Loads a prediction npy file from GCS."""
+        blob = self.client.get_bucket(self.output_bucket).blob(
+            self.get_prediction_npy_path(study_area_id, config_id, chunk_id)
+        )
+        with blob.open("rb") as fd:
+            return np.load(fd)
+
+    def predict_and_save_scenario(
+        self,
+        study_area_id: str,
+        config_id: str,
+        scenario_id: str,
+    ):
+        """Runs predictions for a rainfall scenario and saves outputs to GCS."""
+        parsed_config_id = urllib.parse.quote_plus(config_id)
+        config_dict = (
+            self.db.collection("city_cat_rainfall_configs")
+            .document(parsed_config_id)
+            .get()
+            .to_dict()
+        )
+        rainfall_duration = config_dict["rainfall_duration"]
+        dataset = prediction_dataset.load_prediction_dataset(
+            study_area=study_area_id,
+            city_cat_config=parsed_config_id,
+            batch_size=self.batch_size,
+            storage_client=self.client,
+        )
+        chunk_ids = []
+        for results in self.model.batch_predict_n(dataset, n=rainfall_duration):
+            for result in results:
+                chunk_id = result["chunk_id"]
+                if isinstance(chunk_id, tf.Tensor):
```

Comment:
when isn't chunk_id a tensor? A comment might be helpful here.

Reply:
Added a comment.